collect_app/src/main/java/org/odk/collect/android/utilities/FormListDownloader.java

```diff
@@ -110,213 +80,38 @@ public FormListDownloader(
-                        isNewerFormVersionAvailable, areNewerMediaFilesAvailable));
+        try {
+            ServerFormsDetailsFetcher serverFormsDetailsFetcher = new ServerFormsDetailsFetcher(formRepository, mediaFileRepository, formAPI);
+            List<ServerFormDetails> serverFormDetailsList = serverFormsDetailsFetcher.fetchFormDetails(alwaysCheckMediaFiles);
+            for (ServerFormDetails serverFormDetails : serverFormDetailsList) {
+                formList.put(serverFormDetails.getFormId(), serverFormDetails);
             }
-        } else {
-            // Aggregate 0.9.x mode...
-            // populate HashMap with form names and urls
-            Element formsElement = result.doc.getRootElement();
-            int formsCount = formsElement.getChildCount();
-            String formId = null;
-            for (int i = 0; i < formsCount; ++i) {
-                if (formsElement.getType(i) != Element.ELEMENT) {
-                    // whitespace
-                    continue;
-                }
-                Element child = formsElement.getElement(i);
-                String tag = child.getName();
-                if (tag.equals("formID")) {
-                    formId = XFormParser.getXMLText(child, true);
-                    if (formId != null && formId.length() == 0) {
-                        formId = null;
-                    }
-                }
-                if (tag.equalsIgnoreCase("form")) {
-                    String formName = XFormParser.getXMLText(child, true);
-                    if (formName != null && formName.length() == 0) {
-                        formName = null;
-                    }
-                    String downloadUrl = child.getAttributeValue(null, "url");
-                    downloadUrl = downloadUrl.trim();
-                    if (downloadUrl.length() == 0) {
-                        downloadUrl = null;
-                    }
-                    if (formName == null) {
-                        String error =
-                                "Forms list entry " + Integer.toString(i)
-                                        + " is missing form name or url attribute";
-                        Timber.e("Parsing OpenRosa reply -- %s", error);
-                        formList.clear();
-                        formList.put(
-                                DL_ERROR_MSG,
-                                new FormDetails(application.getString(
-                                        R.string.parse_legacy_formlist_failed, error)));
-                        return formList;
-                    }
-                    formList.put(formName,
-                            new FormDetails(formName, downloadUrl, null, formId, null, null, null, false, false));
-
-                    formId = null;
-                }
+        } catch (FormApiException formApiException) {
+            Timber.e(formApiException);
+
+            switch (formApiException.getType()) {
```

Comment:
I couldn't really come up with a reliable and efficient way to verify that you addressed the exact same exception list when extracting `OpenRosaFormListApi`. That means I'm just trusting you on that and it seems like a potential source of risk.

Reply:
Yeah I agree this is a source of risk!